prefect_aws/filesystems.py

```diff
@@ -0,0 +1,186 @@
+
+from prefect_aws import AwsCredentials, MinIOCredentials
+
+
+class S3Bucket(ReadableFileSystem, WritableFileSystem):
+
+    """
+    Block used to store data using S3-compatible object storage like MinIO.
+
+    Args:
+        bucket_name: Name of your bucket.
+        aws_credentials: A block containing your credentials (choose this
+            or minio_credentials).
+        minio_credentials: A block containing your credentials (choose this
+            or aws_credentials).
+        basepath: Used when you don't want to read/write at base level.
+        endpoint_url: Used for non-AWS configuration. When unspecified,
+            defaults to AWS.
+
+    Example:
+        Load stored S3Bucket configuration:
+        ```python
+        from prefect_aws import S3Bucket
+
+        s3bucket_block = S3Bucket.load("BLOCK_NAME")
+        ```
+    """
+
+    # change
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/uPezmBzEv4moXKdQJ3YyL/a1f029b423cf67f474d1eee33c1463d7/pngwing.com.png?h=250"  # noqa
+    _block_type_name = "S3 Bucket"
+
+    bucket_name: str
+    minio_credentials: Optional[MinIOCredentials]
+    aws_credentials: Optional[AwsCredentials]
+    basepath: Optional[Path]
+    endpoint_url: Optional[str]
+
+    @validator("basepath", pre=True)
+    def cast_pathlib(cls, value):
+
+        """
+        If basepath provided, it means we aren't writing to the root directory
+        of the bucket. We need to ensure that it is a valid path. This is called
+        when the S3Bucket block is instantiated.
+        """
+
+        if isinstance(value, Path):
+            return str(value)
+        return value
+
+    def _resolve_path(self, path: str) -> Path:
+
+        """
+        A helper function used in write_path to join `self.basepath` and `path`.
+
+        Args:
+
+            path: Name of the key, e.g. "file1". Each object in your
+                bucket has a unique key (or key name).
+
+        """
+
+        path = path or str(uuid4())
+
+        # If basepath provided, it means we won't write to the root dir of
+        # the bucket. So we need to add it on the front of the path.
+        path = str(Path(self.basepath) / path) if self.basepath else path
+
+        return path
+
+    def _get_s3_client(self) -> boto3.client:
+
+        """
+        Authenticate MinIO credentials or AWS credentials and return an S3 client.
+        This is a helper function called by read_path() or write_path().
+        """
+
+        s3_client_kwargs = {}
+
+        if self.minio_credentials:
+
+            aws_secret_access_key = self.minio_credentials.minio_root_password
+            s3_client_kwargs.update(
+                aws_access_key_id=self.minio_credentials.minio_root_user,
+                aws_secret_access_key=aws_secret_access_key.get_secret_value(),
+                endpoint_url=self.endpoint_url,
+            )
+
+        s3_client = boto3.client(service_name="s3", **s3_client_kwargs)
+
+        return s3_client
```

Comment:
This method doesn't handle the case where an `AwsCredentials` block is used. This might be a good place to include a check that at either `AwsCredentials` or `MinIOCrednetials` was provided, but not both. If it's possible for that check to be a validator, that might be even better.

Also, both `AwsCredentials` and `MinIOCrednetials` have a `get_boto3_session` method.  I would recommend using those methods here in order to get a session to create your S3 client.

Reply:
That's a great idea!

Reply:
[Add credentials tests on S3Bucket tests](https://github.com/PrefectHQ/prefect-aws/pull/47/commits/d8e8e86269aff6dc585afb11304d40df07bbda1b)
[Add check_credentials() to S3Bucket block](https://github.com/PrefectHQ/prefect-aws/pull/47/commits/9b40284d19c55c6e37c1105d384e01a4240997bb)

Reply:
[Use each credential blocks' method to create s3 client](https://github.com/PrefectHQ/prefect-aws/pull/47/commits/0beec82115200dc82b022e72961381799c1f7e61)